subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt

```diff
@@ -32,6 +39,11 @@ public class IntentionsRouter(
 
         logger.debug("sending intention with id: $intentionId [$intention]")
 
+        while (!isObserving.get()) {
```

Comment:
I see it as an ugly solution, but at least it works.

IDEA test runner with `repeat until failure` gave me stable fails on 2..300 first runs, and 3000+ is green after this fix introduced.

Do you have an idea why it is stuck that way? And maybe there is a better solution?

Reply:
I think it's an expected behavior that sending happens before observing
The key problems there maybe in that `intentionRoutings` is simple mutable map.
Mutable map can't correctly work in multithreaded environment.

You could try to make `send` and `observe` synchronized

Reply:
Yes, you're right!
I used Mutex to synchronize access to map for all methods and it works fine that way